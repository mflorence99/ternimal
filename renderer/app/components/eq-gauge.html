<div class="tag">
  {{ tag }}
</div>

<div (onResize)="handleResize($event)" class="gauge">
  <div
    [ngStyle]="{'background-color': color(ix), 'margin-right.px': params.led.gap, 'width.px': params.led.width}"
    *ngFor="let ix of count | ternimalRange"
    class="led"
  ></div>
</div>

<div class="value">
  {{ value | percent }}
</div>
