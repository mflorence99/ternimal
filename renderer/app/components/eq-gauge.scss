:host {
  align-items: center;
  display: grid;
  font-family: 'B612 Mono', monospace;
  font-size: 0.9rem;
  grid-gap: 0.5rem;
  grid-template-columns: auto 1fr auto;
}

.gauge {
  display: flex;
  height: 100%;
  overflow: hidden;
}

.led {
  flex-shrink: 0;
  height: 100%;
  transition: background-color 0.5s ease-in-out;
}

.tag,
.value {
  width: 2rem;
}

.value {
  text-align: right;
}
