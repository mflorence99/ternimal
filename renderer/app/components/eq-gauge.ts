import { Params } from '../services/params';

import { ChangeDetectionStrategy } from '@angular/core';
import { Component } from '@angular/core';
import { Input } from '@angular/core';
import { ResizeObserverEntry } from 'ngx-resize-observer';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'ternimal-eq-gauge',
  templateUrl: 'eq-gauge.html',
  styleUrls: ['eq-gauge.scss']
})
export class EqGaugeComponent {
  count: number;

  @Input() tag: string;
  @Input() value: number;

  constructor(public params: Params) {}

  color(ix: number): string {
    const ratio = ix / this.count;
    if (ratio > this.value) return 'var(--mat-grey-800)';
    else if (ratio > 0.66) return `var(${this.params.rgb.red})`;
    else if (ratio > 0.33) return `var(${this.params.rgb.yellow})`;
    else return `var(${this.params.rgb.green})`;
  }

  handleResize(resize: ResizeObserverEntry): void {
    this.count = Math.trunc(
      (resize.contentRect.width + this.params.led.gap) /
        (this.params.led.gap + this.params.led.width)
    );
  }
}
