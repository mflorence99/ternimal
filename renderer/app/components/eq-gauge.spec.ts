import { EqGaugeComponent } from './eq-gauge';

import { prepare } from './component.spec';

import { TestBed } from '@angular/core/testing';

import { async } from '@angular/core/testing';

describe('EqGaugeComponent', () => {
  beforeEach(async(() => prepare()));

  test('Component is created', () => {
    const fixture = TestBed.createComponent(EqGaugeComponent);
    const component = fixture.componentInstance;
    expect(component).toBeTruthy();
  });

  test('count is calculated correctly', () => {
    const fixture = TestBed.createComponent(EqGaugeComponent);
    const component = fixture.componentInstance;
    component.handleResize({ contentRect: { width: 80 } } as any);
    expect(component.count).toBe(13);
  });

  test('color is calculated correctly', () => {
    const fixture = TestBed.createComponent(EqGaugeComponent);
    const component = fixture.componentInstance;
    component.handleResize({ contentRect: { width: 80 } } as any);
    expect(component.count).toBe(13);
    component.value = 0.85;
    expect(component.color(0)).toBe(`var(${component.params.rgb.green})`);
    expect(component.color(5)).toBe(`var(${component.params.rgb.yellow})`);
    expect(component.color(10)).toBe(`var(${component.params.rgb.red})`);
    expect(component.color(12)).toBe('var(--mat-grey-800)');
  });
});
