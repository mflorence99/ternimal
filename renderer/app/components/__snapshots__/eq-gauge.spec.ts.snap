// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EqGaugeComponent snapshot 1`] = `
<ternimal-eq-gauge
  __ngContext__={[Function LRootView]}
  params={[Function Params]}
  tag={[Function String]}
  value={[Function Number]}
>
  <figure
    class="eq-gauge monospace"
  >
    <figcaption
      class="tag"
    >
      25%
    </figcaption>
    <ul
      class="gauge"
    >
      
    </ul>
    <figcaption
      class="value"
    >
      25%
    </figcaption>
  </figure>
</ternimal-eq-gauge>
`;
